### PR TITLE
feat(docker): adding dockerized version of server

### PR DESCRIPTION
**Pull request by pascalwhoop**
Adding a docker container for the server-distribution. Further
containers should be added for the weather module and other parts. I am
just starting out contributing so I'll take a while to get a grasp of
the overall structure.

With this, it is only required to have docker installed on the local machine.
Afterwards, executing the `run_docker_image.sh` will create an image and
start a new container based on this image. It will expose the two ports used
by the server and allow the access via `localhost:8080`. Further configurations
should be allowed, so that different variants can be started. I don't quiet understand
the difference yet between the cli and the web interface. Some input would
be appreciated.

Best regards,

### CODE REVIEW

**Comment by jecollins** [comment]
The cli profile runs a single session, while the web2 profile starts a webapp from which multiple sessions can be run. The web2 profile also allows viewing of a running sim as well as viewing of past sims by loading their state logs.

**Comment by jecollins** [comment]
I have not used Docker, and after having read through a brief intro to it I don't understand what it would do for us that maven does not do. Could it be used to run a server and multiple brokers with a single command on a single machine, or on multiple machines? Could it replace Jenkins in the tournament scheduler setup, and if so would there be benefits?

**Comment by pascalwhoop** [comment]
> The cli profile runs a single session, while the web2 profile starts a webapp from which multiple sessions can be run.

Okay thanks, that helps to clarify it for me.

With Docker, it would be a simple task to start up any number of simulations with a selected set of agents that compete in it. This can be useful for agents that require a significant amount of learning like deep RL  agents  such as proximal policy optimization based ones. They have been popular with research into 3D humanoid walking and Atari game solving. 

I want to use this, because I intend to try a proximal policy optimisation based agent in the simulation. Because of the 5s time slot stepping, I intend to automate the start of a few dozen instances in the cloud using the docker container, then let my agent learn on batches of inputs from the simulations. I fear otherwise, an agent that wants to learn to act in the simulation would take a long time to collect enough "training steps", as 5s per step is an awful long time. The policy gradient algorithms are GPU accelerated and would probably mostly idle and wait for a single environment to perform a tick before performing another optimisation step. If I can start e.g. 100 instances in parallel and let it learn in all of them at the same time, it would be a decent speed improvement.

**Comment by jecollins** [comment]
If you hope to use this for running experiments, I've written up [a wiki page on running experiments with a simulation platform like Power TAC](https://github.com/powertac/powertac-server/wiki/Experiments). You might think about how a docker setup might automate more of the process than just running a single session.

**Comment by erikkemperman** [comment]
@pascalwhoop Thanks for the PR!

The web-interface allows you to do mostly the same things as the cli, plus it adds some graphs to monitor the simulations. 

Maybe I am missing something obvious, but can I ask what your main purpose is in wanting to run the server within docker? Is it simply to have an all-in-one setup, to avoid having to install java, maven, git and the server separately, or did you have some specific scenario in mind?

**Comment by pascalwhoop** [comment]
@erikkemperman Yes, firstly it encapsulates all the dependencies into just Docker. There are many Container aa Service offerings from AWS, Azure & Google. Instead of having to create a server manually or using something like [Ansible](https://en.wikipedia.org/wiki/Ansible_(software)), with the Docker setup, you can quickly deploy many simulation servers + competing agents at once and then write an agent that competes in all of them in parallel to increase the learning input for more data hungry learning algorithms. 

Just running a deep NN on one simulation simply won't lead to convergence and running them sequentially takes time. But if you can quickly spin up many at once, it's easier to have multiple environments in which the agent can learn. 

I intend to use [Kubernetes](https://cloud.google.com/kubernetes-engine/) together with work proposed [by TensorFlow](https://github.com/tensorflow/agents) and adapt the library of them to use the powerTAC simulation as the environment instead of the OpenAI Gym. 

The Dockerfile isn't final yet. It's working but it doesn't include any other agents and just runs the web based server. But it's a start. When I come to the point of adding agents as competition etc, I'll extend the file if interest is there for me to share it. 

So: spinning up let's say 100 simulations, connecting my agent to each of them (using a Java-Python bridge with ProtocolBuffers) and letting the agent learn in all of them.

**Comment by pascalwhoop** [comment]
I recognize that my contribution doesn't seem to be obvious so I don't mind if you close this and once I managed to show some results in my thesis, I can say Hi again, with an exact description of how I use it. I just know that many in the Community appreciate when something is also available as a docker image as it is easy to try something without spamming your system with dependencies you usually never use. Not saying Java & `mvn` aren't essential tools for a Java Developer, but for someone from the Python space it's foreign, just as `pip` is for a Java dev or `npm` is for someone not versed in NodeJS.

**Comment by erikkemperman** [comment]
@pascalwhoop Thanks for elaborating, that sounds really interesting! I think we should adopt this, once you feel it's ready for general consumption. Only slight discomfort for me is that, as far as I know, the core developers are not (yet?) particularly knowledgeable about docker and might need to learn more about it before we commit to maintaining this... @powertac/owners Thoughts?

**Comment by jecollins** [comment]
Pascal, does your agent need to interact with a number of running instances in order to learn, or does it just need to capture the message flow. If it's the latter, you should not need to run simulations; it should be enough to suck up a big stack of logs, for example the logs from the 2017 competition. We have a tool that reads state logs and turns them back into Java instances, while keeping simulation time. That's what the visualizer does to view old games. If you want to get that data as a stream of messages such as a broker would see, that should be easy to implement. If you want to actually interact with a number of running instances in parallel, that could be a fairly significant performance challenge.

On the other hand, it's possible that our tournament and experiment schedulers could use docker and possibly kubernetes rather than jenkins to manage their workflows. So that could be very interesting. Also, your use case sounds much better suited to the cli profile rather than the web profile.

**Comment by pascalwhoop** [comment]
Since it is a RF agent it would need to perform actions and observe the environments results, hence it needs an *alive* environment which reacts to its actions. 

I have realised I can do a lot more if I take some time so I will close this PR and create such a setup on my own account, developing a more elaborate deployment possibility with Kubernetes configurations etc. Once I think it is ready, I will notify the team and see if you like it. I also have the intention of a python bridge, to allow writing an agent in Python. But I will make another issue for this to facilitate a possible discussion what is desired by the team.
